sophia/src/triple/stream/_resolve.rs

```diff
@@ -0,0 +1,140 @@
+        E: std::error::Error,
+    {
+        let source = &mut self.source;
+        let base = &self.base;
+
+        source
+            .try_for_some_triple(&mut |t| match base.resolve(&t) {
+                Ok(resolved) => f(StreamedTriple::by_value(resolved)).map_err(SinkError),
+                Err(e) => Err(SourceError(e)),
+            })
+            .map_err(|e| match e {
+                SourceError(e) => SourceError(SourceError(e)),
+                SinkError(SourceError(e)) => SourceError(SinkError(e)),
+                SinkError(SinkError(e)) => SinkError(e),
+            })
+    }
+}
+
+impl<'a, S> QuadSource for Resolver<'a, S>
+where
+    S: QuadSource,
+    QSData<S>: From<String>,
+{
+    type Error = StreamError<S::Error, TermError>;
+    type Quad = QByValue<TupleQuad<QSData<S>>>;
+
+    /// This is an adapter that might fail. Accordingly, there are three error
+    /// sources. The source `QuadSource`, resolving the quads and the
+    /// closure. Errors are propagated as follows:
+    ///
+    /// - Source - Source => From original source.
+    /// - Source - Sink   => From resolving IRIs.
+    /// - Sink => From closure.
+    fn try_for_some_quad<F, E>(&mut self, f: &mut F) -> StreamResult<bool, Self::Error, E>
+    where
+        F: FnMut(StreamedQuad<Self::Quad>) -> Result<(), E>,
+        E: std::error::Error,
+    {
+        let source = &mut self.source;
+        let base = &self.base;
+
+        source
+            .try_for_some_quad(&mut |q| match base.resolve(&q) {
+                Ok(resolved) => f(StreamedQuad::by_value(resolved)).map_err(SinkError),
+                Err(e) => Err(SourceError(e)),
+            })
+            .map_err(|e| match e {
+                SourceError(e) => SourceError(SourceError(e)),
+                SinkError(SourceError(e)) => SourceError(SinkError(e)),
+                SinkError(SinkError(e)) => SinkError(e),
+            })
+    }
+}
+
+impl<'a, S> IntoIterator for Resolver<'a, S>
+where
+    S: TripleSource,
+    TSData<S>: From<String>,
+{
+    type Item = StreamResult<[Term<TSData<S>>; 3], S::Error, TermError>;
+    type IntoIter = ResolveIter<'a, S>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        ResolveIter {
+            base: self.base,
+            source: self.source,
+        }
+    }
+}
+
+pub struct ResolveIter<'a, S> {
+    base: IriParsed<'a>,
+    source: S,
+}
+
+impl<'a, S> Iterator for ResolveIter<'a, S>
+where
+    S: TripleSource,
+    TSData<S>: From<String>,
+{
+    type Item = StreamResult<[Term<TSData<S>>; 3], S::Error, TermError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut triple = None;
+        let source = &mut self.source;
+        let base = &self.base;
+
+        match source.for_some_triple(&mut |t| {
+            triple = Some(base.resolve(&t).map_err(SinkError));
+        }) {
+            Ok(true) => triple,
+            Ok(false) => None,
+            Err(e) => Some(Err(SourceError(e))),
+        }
+    }
```

Comment:
This may not always work: try_for_some_triple may invoke the function more than once... So the only safe way to convert a Triple- or QuadSource to an iterator is to use an internal buffer (see triple::stream::_iterator for an example).

Reply:
Removed in efe2a62